src/server/gc_worker/compaction_filter.rs

```diff
@@ -232,41 +237,38 @@ impl CompactionFilter for WriteCompactionFilter {
             self.key_prefix.extend_from_slice(key_prefix);
             self.remove_older = false;
             self.switch_key_metrics();
-            // The level's tail delete mark can be removed because
-            // the key is not the last one in the level.
-            self.leveled_tail_deletes.remove(&level);
         }
 
-        self.versions += 1;
-        if commit_ts > safe_point {
+        if commit_ts > self.safe_point {
             return false;
         }
 
+        self.versions += 1;
         let mut filtered = self.remove_older;
-        let WriteRef {
-            write_type,
-            start_ts,
-            short_value,
-        } = WriteRef::parse(value).unwrap();
+        let write = match WriteRef::parse(value) {
+            Ok(write) => write,
+            // Invalid MVCC keys, don't touch them.
+            Err(_) => return false,
+        };
+
         if !self.remove_older {
             // here `filtered` must be false.
-            match write_type {
+            match write.write_type {
                 WriteType::Rollback | WriteType::Lock => filtered = true,
+                WriteType::Put => self.remove_older = true,
                 WriteType::Delete => {
                     self.remove_older = true;
                     if self.bottommost_level {
-                        // Handle delete marks if the bottommost level is reached.
                         filtered = true;
-                        self.leveled_tail_deletes.insert(level, key.to_vec());
+                        self.gc_before_delete_mark(key, key_prefix);
```

Comment:
Can we promise that delete operations of these old versions are flushed before the current `delete` record is filtered out?

Reply:
All old versions are deleted in `Drop`. This will be called before the new generated SST by the compaction has been insert into manifest.